### ISSUE TITLE
TypeError when using Bookshelf.js

### ISSUE THREAD

**Issue by owm111** [nuxt/nuxt.js#1747]
I have been trying to get Vuex (inside of Nuxt.js) to work with a SQLite database using Bookshelf.js and Knex.js. I have been able to get Bookshelf to function outside of Nuxt, but when I put the same code inside of Nuxt, I get this in my browser console upon running a model:
```
bluebird.js:1542
Unhandled rejection TypeError: _this.driver.Database is not a constructor
    acquireRawConnection/<@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:38103:16
    acquireRawConnection@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:38102:12
    create@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:32558:9
    _createResource@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:29271:3
    dispense@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:29231:5
    acquire@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:29353:3
    acquireConnection/<@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:32608:7
From previous event:
    acquireConnection@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:32599:12
    ensureConnection/</<@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:42442:16
From previous event:
    ensureConnection/<@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:42440:35
From previous event:
    ensureConnection@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:42439:12
    run@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:42285:37
    ["./node_modules/knex/lib/interface.js"]/exports.default/Target.prototype.then@http://localhost:3000/_nuxt/pages/testing.e359f91ab7a42b76ee09.js:39654:18
```
I am using the same setup I used outside of Nuxt, but now it isn't working. This makes me think that the issue is on Nuxt's side.

P.S. If I am doing anything wrong or if there is a better way to achieve this (have Vuex interact with a database), please tell me

-----

To reproduce error (may upload repo in the morning; getting late):
1. Make a new Nuxt project
```sh
vue init nuxt/starter bug-test
cd bug-test
npm install
```
2. Install Bookshelf, etc.
```sh
npm install bookshelf knex
npm install sqlite3@3.1.11 --save-exact # 3.1.12 causes bug which makes it impossible to `npm install`
```
3. Add database folder (for organization)
```sh
mkdir database
cd database
```
4. Add an empty SQLite database
```sh
sqlite3 db.sqlite
# SQLite version 3.20.1...
sqlite> create table blank; # not adding columns causes it to fail but still write a file; not best method but works
sqlite> .exit
```
5. Add Bookshelf.js files (add these to `database/` directory; same as db.sqlite)
    1. knexfile.js
    ```js
    module.exports = {
      client: 'sqlite3',
      connection: {
        filename: 'db.sqlite'
      },
      useNullAsDefault: true
    }
    ```
    2. bookshelf.js
     ```js
    let knex = require('knex')(require('./knexfile'))

    let bookshelf = require('bookshelf')(knex)

    module.exports = bookshelf
    ```
   3. migrations/users.js
   ```js
    exports.up = function (knex) {
      return knex.schema
        .createTable('users', table => {
          table.increments('id').primary()
          table.string('firstName')
          table.string('lastName')
          table.string('emailAddress')
        })
    }

    exports.down = function (knex) {
      return knex.schema
        .dropTable('users')
    }
    ```
    4. models/users.js
     ```js
    var bookshelf = require('../bookshelf')

    var Users = bookshelf.Model.extend({
      tableName: 'users'
    })

    module.exports = Users
    ```
6. Add new page
```vue
<template lang="html">
  <section>
    <h1>Testing Page</h1>
    <pre>{{users}}</pre>
  </section>
</template>

<script>
import Users from '~/database/models/users'

export default {
  computed: {
    users () {
      let users
      Users.fetchAll().then(function (contact) { users = contact.toJSON() })
      return users
    }
  }
}
</script>
```
7. Migrate and start it up
```sh
cd ../ # if still in database/ or pages/
./node_modules/.bin/knex migrate:latest --knexfile database/knexfile.js
npm run dev
```
8. Open http://localhost:3000/testing (or whatever you called your page) in a browser and open console

-----

Any help/feedback appreciated

**Comment by vuchl** [comment]
are you sure you can use bookshelf on the client side?

**Comment by owm111** [comment]
I have been messing around with this since posting this issue.

So far, I have found a way to get the data from the database. What I did differently was:
- Creating the database by making a new blank file instead of using sqlite3 (I don't know if this actually changes anything, but it works now)
- Changing the `connection.filename` in knexfile.js to `database/db.sqlite` (I'm pretty sure this is the thing that makes it work)
- **Instead of using computed property, use the asyncData method**. Require the model, fetch data and return it or commit to Vuex state. Since (to my understanding) asyncData is always done on the server, Bookshelf always has access to the database file (to @vuchl 's point). I put this into my default layout, it saves it to a property and to the Vuex state using a mutation (also works in a page, too, but I think a layout is easiest):
```js
// ...
  async asyncData ({store}) {
    let Contact = require('~/database/models/users')
    let users = await Contact.fetchAll()
    store.commit('setTest', users.toJSON())
    return {users}
  }
// ...
```
I am beginning to mess around with updating the database when the page is loaded (probably by looking for differences in the Vuex state and fetched data and inserting/dropping/updating the differences).
